h2o-persist-hdfs/src/main/java/water/persist/security/HdfsDelegationTokenRefresher.java

```diff
@@ -0,0 +1,215 @@
+        FileUtils.makeSureDirExists(tmpDir);
+        File keytabFile = new File(tmpDir, "hdfs_auth_keytab");
+        byte[] byteArr = BinaryFileTransfer.convertStringToByteArr(authKeytab);
+        BinaryFileTransfer.writeBinaryFile(keytabFile.getAbsolutePath(), byteArr);
+        return keytabFile.getAbsolutePath();
+    }
+
+    private final ScheduledExecutorService _executor = Executors.newSingleThreadScheduledExecutor(
+            new ThreadFactoryBuilder().setDaemon(true).setNameFormat("hdfs-token-refresher-%d").build()
+    );
+
+    private final String _authPrincipal;
+    private final String _authKeytabPath;
+    private final String _authUser;
+    private final double _intervalRatio;
+
+    public HdfsDelegationTokenRefresher(
+            String authPrincipal,
+            String authKeytabPath,
+            String authUser,
+            double intervalRatio
+    ) {
+        _authPrincipal = authPrincipal;
+        _authKeytabPath = authKeytabPath;
+        _authUser = authUser;
+        _intervalRatio = intervalRatio;
+    }
+
+    public void start() {
+        long interval;
+        try {
+             interval = getTokenRenewalInterval(loginAuthUser());
+        } catch (IOException | InterruptedException e) {
+            log("Failed to determine delegation token renewal interval", e);
+            return;
+        }
+        final long actualInterval;
+        if (interval == Long.MAX_VALUE) {
+            log("Couldn't determine tokens renewal interval, will use 12h", null);
+            actualInterval = 12 * 3600 * 1000; // 12h
+        } else
+            actualInterval = (long) (interval * _intervalRatio);
+        log("Determined token renewal interval = " + interval + "ms. Using actual interval = " + actualInterval + "ms.", null);
+        _executor.scheduleAtFixedRate(this, 0, interval, TimeUnit.MILLISECONDS);
+    }
+
+    private static void log(String s, Exception e) {
+        System.out.println("TOKEN REFRESH: " + s);
+        if (e != null) {
+            e.printStackTrace(System.out);
+        }
+    }
+
+    @Override
+    public void run() {
+        if (Paxos._cloudLocked && !(H2O.CLOUD.leader() == H2O.SELF)) {
+            // cloud is formed the leader will take of subsequent refreshes
+            _executor.shutdown();
+            return;
+        }
+        try {
+            Credentials creds = refreshTokens(loginAuthUser());
+            distribute(creds);
+        } catch (IOException | InterruptedException e) {
+            log("Failed to refresh token.", e);
+        }
+    }
+
+    private Credentials refreshTokens(UserGroupInformation tokenUser) throws IOException, InterruptedException {
+        return tokenUser.doAs((PrivilegedExceptionAction<Credentials>) () -> {
+            Credentials creds = new Credentials();
+            try {
+                Token<?>[] tokens = fetchDelegationTokens(_authUser, creds);
```

Comment:
qq: where is the renewing done? Does it still needs to b implemented?

Reply:
this call will always retrieve a fresh set of tokens, regardless if there are existing tokens around because I am passing an empty Credentials object

if I used the actual users Credentials I could get tokens that are about to expire